common/transport/mqtt/src/mqtt_base.ts

```diff
@@ -6,41 +6,83 @@
 import { EventEmitter } from 'events';
 import * as machina from 'machina';
 import { Client as MqttClient, IClientOptions, IClientPublishOptions, IClientSubscribeOptions } from 'mqtt';
+import * as uuid from 'uuid';
 import * as dbg from 'debug';
 const debug = dbg('azure-iot-mqtt-base:MqttBase');
 import { errors, results, SharedAccessSignature, X509 } from 'azure-iot-common';
 
-/*Codes_SRS_NODE_COMMON_MQTT_BASE_16_004: [The `MqttBase` constructor shall instanciate the default MQTT.JS library if no argument is passed to it.]*/
+
+class OnTheWirePublish {
+  enqueuedTimeSecondsSinceEpoch: Number;
+  topic: string;
+  payload: any;
+  options: IClientPublishOptions;
+  publishCallback: (err?: Error, result?: any) => void;
+  identifier: string;
+  constructor(enqueuedTimeSecondsSinceEpoch: Number,topic: string, payload: any, options: IClientPublishOptions, publishCallback: (err?: Error, result?: any) => void, identifier: string) {
+    this.enqueuedTimeSecondsSinceEpoch = enqueuedTimeSecondsSinceEpoch;
+    this.topic = topic;
+    this.payload = payload;
+    this.options = options;
+    this.publishCallback = publishCallback;
+    this.identifier = identifier;
+  }
+}
+
+
+/*Codes_SRS_NODE_COMMON_MQTT_BASE_16_004: [The `MqttBase` constructor shall instantiate the default MQTT.JS library if no argument is passed to it.]*/
 /*Codes_SRS_NODE_COMMON_MQTT_BASE_16_005: [The `MqttBase` constructor shall use the object passed as argument instead of the default MQTT.JS library if it's not falsy.]*/
 /**
  * @private
  */
 export class MqttBase extends EventEmitter {
-  private mqttprovider: any;
+  private mqttProvider: any;
   private _config: MqttBaseTransportConfig;
   private _mqttClient: MqttClient;
   private _fsm: any;
   private _options: any;
+  private _publishesOnTheWire: { [uuidKey: string]: OnTheWirePublish };
 
-  constructor(mqttprovider?: any) {
+  constructor(mqttProvider?: any) {
     super();
-    this.mqttprovider = mqttprovider ? mqttprovider : require('mqtt');
+    this.mqttProvider = mqttProvider ? mqttProvider : require('mqtt');
+    this._publishesOnTheWire = {};
 
     this._fsm = new machina.Fsm({
       namespace: 'mqtt-base',
       initialState: 'disconnected',
       states: {
         disconnected: {
           _onEnter: (callback, err) => {
-            if (this._mqttClient) {
-              this._mqttClient.removeAllListeners();
-              this._mqttClient = undefined;
-            }
-
+            debug('In MQTT base FSM - entered onEnter for disconnect');
+            //
+            // The semantics of this _onEnter for the disconnected state (which is the initial state)
+            // is that we got here from another one of the states of this FSM.
+            //
+            // As the initial state, the _onEnter is not invoked when this FSM is initially used.
```

Comment:
wait, is this true?

Reply:
I certainly thought so.

---
In reply to: [431342528](https://github.com/Azure/azure-iot-sdk-node/pull/808#discussion_r431342528) [](ancestors = 431342528)

Reply:
And I would be wrong.  Thank you Yoseph.

---
In reply to: [431362075](https://github.com/Azure/azure-iot-sdk-node/pull/808#discussion_r431362075) [](ancestors = 431362075,431342528)